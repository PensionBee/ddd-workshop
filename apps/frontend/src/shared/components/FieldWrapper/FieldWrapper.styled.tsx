import { Box } from "@shared/components/Box";
import { theme } from "@shared/styling/theme";
import styled, { css } from "styled-components";

export type StyledFieldWrapperProps = {
  $isError?: boolean;
  $isDisabled?: boolean;
  $hasValue?: boolean;
  $hasFocusState?: boolean;
};

export const StyledFieldWrapper = styled.span<StyledFieldWrapperProps>`
  position: relative;
  flex: 1;
  display: flex;
  flex-direction: row;
  flex-wrap: "nowrap";
  border: 2px solid ${() => theme.colors.grey3};
  background-color: ${() => theme.colors.grey8};
  border-radius: 4px;
  transition: border 0.2s ease, box-shadow 0.2s ease;
  &:hover {
    border: 2px solid ${() => theme.colors.grey2};
  }
  ${({ $isError }) =>
    $isError &&
    css`
      &,
      &:hover {
        border: 2px solid ${() => theme.colors.red100};
      }
    `}
  ${({ $isDisabled }) =>
    $isDisabled &&
    css`
      cursor: not-allowed;
      opacity: 0.5;
      &:hover {
        border: 2px solid ${() => theme.colors.grey3};
      }
    `}
  ${({ $hasValue }) =>
    $hasValue &&
    css`
      &,
      &:hover {
        border: 2px solid ${() => theme.colors.grey2};
      }
    `}
  ${({ $hasFocusState }) =>
    $hasFocusState &&
    css`
      &:focus-within {
        outline: 3px solid ${() => theme.colors.blue100};
      }
    `}
`;

export const StyledChildWrapper = styled(Box)``;
