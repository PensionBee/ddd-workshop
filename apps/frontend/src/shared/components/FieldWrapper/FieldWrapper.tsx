import { StyledChildWrapper, StyledFieldWrapper } from "./FieldWrapper.styled";
import { Prefix, type PrefixProps } from "./Prefix/Prefix";
import { Suffix, type SuffixProps } from "./Suffix/Suffix";

type BaseFieldWrapperProps = {
  /** The input to wrap */
  children?: React.ReactNode;
  /** The error message to display for the input */
  error?: string;
  /** Whether the input is disabled */
  disabled?: boolean;
  /** Whether the input has a value */
  hasValue?: boolean;
  /** Whether the field has a focus state */
  hasFocusState?: boolean;
};

export type FieldWrapperProps = BaseFieldWrapperProps &
  PrefixProps &
  SuffixProps;

export const FieldWrapper: React.FC<FieldWrapperProps> = ({
  children,
  error,
  disabled,
  hasValue,
  hasFocusState = true,
  prefixIcon,
  prefixIconFill,
  prefixText,
  suffixIcon,
  suffixIconFill,
  suffixText,
  ...props
}) => {
  const prefixProps = {
    prefixIcon,
    prefixIconFill,
    prefixText,
  } satisfies PrefixProps;

  const suffixProps = {
    suffixIcon,
    suffixIconFill,
    suffixText,
  } satisfies SuffixProps;

  return (
    <StyledFieldWrapper
      $isError={Boolean(error)}
      $isDisabled={Boolean(disabled)}
      $hasValue={Boolean(hasValue)}
      $hasFocusState={hasFocusState}
      {...props}
    >
      <Prefix {...prefixProps} />
      <StyledChildWrapper flex="1" d="flex" nowrap>
        {children}
      </StyledChildWrapper>
      <Suffix {...suffixProps} />
    </StyledFieldWrapper>
  );
};
