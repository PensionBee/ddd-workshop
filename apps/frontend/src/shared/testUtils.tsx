import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
