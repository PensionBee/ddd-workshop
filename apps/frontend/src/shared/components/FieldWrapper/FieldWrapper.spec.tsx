import { render, vi } from "@shared/testUtils";
import { FieldWrapper } from "./FieldWrapper";

const INPUT_CONTENT = "Input content";
const PREFIX_TEXT = "Prefix text";
const PREFIX_ICON = "AlertInfo";
const SUFFIX_TEXT = "Suffix text";

describe("FieldWrapper", () => {
  describe("With prefix", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <FieldWrapper prefixText={PREFIX_TEXT}>
          <div>{INPUT_CONTENT}</div>
        </FieldWrapper>
      );

      expect(getByText(PREFIX_TEXT)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("With prefix icon", () => {
    it("should render correctly", () => {
      const { container, asFragment } = render(
        <FieldWrapper prefixIcon={PREFIX_ICON}>
          <div>{INPUT_CONTENT}</div>
        </FieldWrapper>
      );
      const svgContainer = container.querySelector(
        `[data-rc="SmallSystemIcon.${PREFIX_ICON}"]`
      );

      expect(svgContainer).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("With suffix", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <FieldWrapper suffixText={SUFFIX_TEXT}>
          <div>{INPUT_CONTENT}</div>
        </FieldWrapper>
      );

      expect(getByText(SUFFIX_TEXT)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render correctly with suffix icon", () => {
      const { container, asFragment } = render(
        <FieldWrapper suffixIcon={PREFIX_ICON}>
          <div>{INPUT_CONTENT}</div>
        </FieldWrapper>
      );
      const svgContainer = container.querySelector(
        `[data-rc="SmallSystemIcon.${PREFIX_ICON}"]`
      );

      expect(svgContainer).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
