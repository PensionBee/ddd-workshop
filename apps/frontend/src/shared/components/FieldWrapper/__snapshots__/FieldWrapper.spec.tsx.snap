// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FieldWrapper > With prefix > should render correctly 1`] = `
<DocumentFragment>
  <span
    class="sc-gsnScm bgFqSI"
  >
    <div
      class="sc-bczSft eblhFG"
    >
      <span
        class="sc-eCYdKt hUJlhy"
        data-rc="BodySmallB"
      >
        Prefix text
      </span>
    </div>
    <div
      class="sc-bczSft ghPjgr sc-dkzCjb"
    >
      <div>
        Input content
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`FieldWrapper > With suffix > should render correctly 1`] = `
<DocumentFragment>
  <span
    class="sc-gsnScm bgFqSI"
  >
    <div
      class="sc-bczSft ghPjgr sc-dkzCjb"
    >
      <div>
        Input content
      </div>
    </div>
    <div
      class="sc-bczSft fNJJm"
    >
      <span
        class="sc-eCYdKt hUJlhy"
        data-rc="BodySmallB"
      >
        Suffix text
      </span>
    </div>
  </span>
</DocumentFragment>
`;

exports[`FieldWrapper With prefix icon should render correctly 1`] = `
<DocumentFragment>
  <span
    class="sc-gswNZR cvCIbd"
  >
    <div
      class="sc-bcXHqe ctGgCZ"
    >
      <span
        class="sc-hLBbgP ktrFWh"
        data-rc="SmallSystemIcon.AlertInfo"
      >
        <svg>
          alert-info.svg
        </svg>
      </span>
    </div>
    <div
      class="sc-bcXHqe grQKHY sc-dkrFOg"
    >
      <div>
        Input content
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`FieldWrapper With prefix should render correctly 1`] = `
<DocumentFragment>
  <span
    class="sc-gswNZR cvCIbd"
  >
    <div
      class="sc-bcXHqe ctGgCZ"
    >
      <span
        class="sc-eDvSVe enytKW"
        data-rc="BodySmallB"
      >
        Prefix text
      </span>
    </div>
    <div
      class="sc-bcXHqe grQKHY sc-dkrFOg"
    >
      <div>
        Input content
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`FieldWrapper With suffix should render correctly 1`] = `
<DocumentFragment>
  <span
    class="sc-gswNZR cvCIbd"
  >
    <div
      class="sc-bcXHqe grQKHY sc-dkrFOg"
    >
      <div>
        Input content
      </div>
    </div>
    <div
      class="sc-bcXHqe HcrbJ"
    >
      <span
        class="sc-eDvSVe enytKW"
        data-rc="BodySmallB"
      >
        Suffix text
      </span>
    </div>
  </span>
</DocumentFragment>
`;

exports[`FieldWrapper With suffix should render correctly with suffix icon 1`] = `
<DocumentFragment>
  <span
    class="sc-gswNZR cvCIbd"
  >
    <div
      class="sc-bcXHqe grQKHY sc-dkrFOg"
    >
      <div>
        Input content
      </div>
    </div>
    <div
      class="sc-bcXHqe HcrbJ"
    >
      <span
        class="sc-hLBbgP ktrFWh"
        data-rc="SmallSystemIcon.AlertInfo"
      >
        <svg>
          alert-info.svg
        </svg>
      </span>
    </div>
  </span>
</DocumentFragment>
`;
